packages/mdc-menu/simple/foundation.js

```diff
@@ -399,69 +335,202 @@ class MDCSimpleMenuFoundation extends MDCFoundation {
   }
 
-  /** @private */
-  autoPosition_() {
-    if (!this.adapter_.hasAnchor()) {
-      return;
-    }
+  /**
+   * @return {AutoLayoutMeasures} Measurements used to position menu popup.
+   */
+  getAutoLayoutMeasures_() {
+    const anchorRect = this.adapter_.getAnchorDimensions();
+    const viewport = this.adapter_.getWindowDimensions();
+
+    return {
+      anchorRect: anchorRect,
+      viewport: viewport,
+      viewportMargin: {
+        top: anchorRect.top,
+        right: viewport.width - anchorRect.right,
+        left: anchorRect.left,
+        bottom: viewport.height - anchorRect.bottom,
+      },
+      anchorHeight: anchorRect.height,
+      anchorWidth: anchorRect.width,
+      menuHeight: this.dimensions_.height,
+      menuWidth: this.dimensions_.width,
+    };
+  }
 
+  /**
+   * Computes the corner of the anchor from which to animate and position the menu.
+   * @return {Corner}
+   * @private
+   */
+  getOriginCorner_() {
     // Defaults: open from the top left.
-    let vertical = 'top';
-    let horizontal = 'left';
-
-    const anchor = this.adapter_.getAnchorDimensions();
-    const windowDimensions = this.adapter_.getWindowDimensions();
+    let corner = Corner.TOP_LEFT;
+
+    const {anchorRect, viewportMargin, anchorHeight, anchorWidth, menuHeight, menuWidth} = this.measures_;
+    const bottomAligned = this.anchorCorner_ & CornerBit.BOTTOM;
+    const availableTop = bottomAligned ? anchorRect.top + anchorHeight + this.anchorMargin_.bottom
+      : viewportMargin.top + this.anchorMargin_.top;
+    const availableBottom = bottomAligned ? viewportMargin.bottom - this.anchorMargin_.bottom
+      : viewportMargin.bottom + anchorHeight - this.anchorMargin_.bottom;
+
+    const topOverflow = menuHeight - availableTop;
+    const bottomOverflow = menuHeight - availableBottom;
+    if (bottomOverflow > 0 && topOverflow < bottomOverflow) {
+      corner |= CornerBit.BOTTOM;
+    }
 
-    const topOverflow = anchor.top + this.dimensions_.height - windowDimensions.height;
-    const bottomOverflow = this.dimensions_.height - anchor.bottom;
-    const extendsBeyondTopBounds = topOverflow > 0;
+    const leftAligned = this.adapter_.isRtl() ? (this.anchorCorner_ & (CornerBit.FLIP_RTL | CornerBit.RIGHT)) :
+      !(this.anchorCorner_ & CornerBit.RIGHT);
+    const availableLeft = leftAligned ? viewportMargin.left + this.anchorMargin_.left :
+      viewportMargin.left + anchorWidth + this.anchorMargin_.right;
+    const availableRight = leftAligned ? anchorWidth - this.anchorMargin_.left + viewportMargin.right :
+      viewportMargin.right - this.anchorMargin_.right;
 
-    if (extendsBeyondTopBounds) {
-      if (bottomOverflow < topOverflow) {
-        vertical = 'bottom';
+    const leftOverflow = menuWidth - availableLeft;
+    const rightOverflow = menuWidth - availableRight;
+    if (this.adapter_.isRtl()) {
+      if (leftOverflow < 0 && this.anchorCorner_ & CornerBit.FLIP_RTL & ~CornerBit.RIGHT) {
```

Comment:
There's no else to this inner `if`? Can this simply be flattened into the `if` above?

Reply:
because of the else if below....

Reply:
Actually, can't this be represented a bit more concisely like so, rather than having multiple variably-nested branches that result in the same action?

```js
const shouldUseRightCorner =
  (this.adapter_.isRtl() && leftOverflow < 0 && this.anchorCorner_ & CornerBit.FLIP_RTL & ~CornerBit.RIGHT) ||
  (!this.adapter_.isRtl() && rightOverflow > 0 && leftOverflow < rightOverflow);
if (shouldUseRightCorner) {
  corner |= CornerBit.RIGHT;
}
```